sdks/java/core/src/main/java/org/apache/beam/sdk/util/WatermarkHold.java

```diff
@@ -328,23 +334,37 @@ private Instant addGarbageCollectionHold(ReduceFn<?, ?, ?, W>.Context context) {
     // at garbage collection time, and garbage collection time is strictly after the
     // end of window. (All non-empty panes will have holds at their output
     // time derived from their incoming elements and no additional hold is required.)
-    if (context.windowingStrategy().getClosingBehavior() == ClosingBehavior.FIRE_ALWAYS
-        && windowingStrategy.getAllowedLateness().isLongerThan(Duration.ZERO)) {
-      Instant gcHold = context.window().maxTimestamp().plus(windowingStrategy.getAllowedLateness());
-      Instant outputWM = timerInternals.currentOutputWatermarkTime();
-      Instant inputWM = timerInternals.currentInputWatermarkTime();
-
-      WindowTracing.trace(
-          "WatermarkHold.addGarbageCollectionHold: garbage collection at {} hold for "
-          + "key:{}; window:{}; inputWatermark:{}; outputWatermark:{}",
-          gcHold, context.key(), context.window(), inputWM, outputWM);
+    Instant outputWM = timerInternals.currentOutputWatermarkTime();
+    Instant inputWM = timerInternals.currentInputWatermarkTime();
+
+    String which;
+    boolean added;
+    Instant eow = context.window().maxTimestamp();
+    Instant gcHold = eow.plus(windowingStrategy.getAllowedLateness());
+    if (context.windowingStrategy().getClosingBehavior() != ClosingBehavior.FIRE_ALWAYS) {
+      which = "unnecessary (not FIRE_ALWAYS)";
+      added = false;
+    } else if (!windowingStrategy.getAllowedLateness().isLongerThan(Duration.ZERO)) {
+      which = "unnecessary (no allowed lateness)";
+      added  = false;
+    } else if (!eow.isBefore(GlobalWindow.INSTANCE.maxTimestamp())) {
```

Comment:
If a runner does choose to advance the watermark to the max, the buffered data should be emitted at the EOW timestamp, even if in global window. Does other code compensate for the lack of a hold here?